docs(deployment): add cost-benefit analysis with different options (#6026)
facebook/docusaurus: website/docs/deployment.mdx

:::

## Choosing a hosting provider

There are a few common hosting options:

- [Self hosting](#self-hosting) with an HTTP server like Apache2 or Nginx;

> - Jamstack providers, e.g. [Netlify](#deploying-to-netlify) and [Vercel](#deploying-to-vercel). We will use Netlify as reference, but the same reasoning can apply to other providers.

Reviewer: I like the Netlify DX but I don't think it's fair to use it as a primary example considering Vercel has 6000 free build minutes for hobby projects.

Author: I'm not familiar with Vercel and I'm slightly bugged by them commenting on commits (at least what I observed a lot with other projects; pretty sure it's configurable). Anyways I just changed it to `We will use them`

> - [GitHub pages](#deploying-to-github-pages). (By definition, it is also Jamstack, but we compare it separately.)

Reviewer 2: ```suggestion
- [GitHub Pages](#deploying-to-github-pages). (By definition, it is also Jamstack, but we compare it separately.)
```

If you are unsure of which one to choose, ask the following questions:

<details>

<summary>
  How much resource (person-hours, money) am I willing to invest in this?
</summary>

- 🔴 Self-hosting is the hardest to set up—you would usually need an experienced person to manage this. Cloud services is almost never free, and setting up an on-site server and connecting it to the WAN can be more costly.

> - 🟢 Jamstack providers can help you set up a working website in almost no time and offers features like server-side redirects that are easily configurable. Build time can be limited—you would need to pay in order to get more build time quota. Check the pricing page of your provider for details.

Reviewer: Worth mentioning that those are more powerful, and they offer generous free plans for hobby projects, and are not very expensive for companies

> - 🟡 The GitHub pages deployment workflow can be tedious to set up. (Evidence: compare the length of [Deploying to Netlify](#deploying-to-netlify) and [Deploying to GitHub pages](#deploying-to-github-pages)!) However, this service (including build and deployment) is always free, and we have detailed instructions to help you make it work.

Reviewer 2: ```suggestion
- 🟡 The GitHub Pages deployment workflow can be tedious to set up. (Evidence: compare the length of [Deploying to Netlify](#deploying-to-netlify) and [Deploying to GitHub Pages](#deploying-to-github-pages)!) However, this service (including build and deployment) is always free, and we have detailed instructions to help you make it work.
```

> - 🟡 The GitHub Pages deployment workflow can be tedious to set up. (Evidence: compare the length of [Deploying to Netlify](#deploying-to-netlify) and [Deploying to GitHub Pages](#deploying-to-github-pages)!) However, this service (including build and deployment) is always free, and we have detailed instructions to help you make it work.

Reviewer: maybe just mention the length got the GH instructions, not Netlify?

Also it's not 100% free either, it's free if you run deploy locally manually, not if you automate with GH actions

I would mostly mention that GH pages is a jamstack provider that gives you very limited server options

Author: GitHub actions is entirely free. Public repos have unlimited build time: https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions

> GitHub Actions usage is free for both public repositories and self-hosted runners. For private repositories, each GitHub account receives a certain amount of free minutes and storage,

</details>

<details>

<summary>How much server-side configuration would I need?</summary>

- 🟢 With self-hosting, you have access to the entire server's configuration. You can configure the virtual host to serve different content based on the request URL; you can do complicated server-side redirects; you can put part of the site behind authentication... If you need a lot of server-side features, self-host your website.
- 🟡 Jamstack usually offers some server-side configurations, e.g. URLs formatting (trailing slashes), server-side redirects...
- 🔴 GitHub pages doesn't expose server-side configurations besides enforcing HTTPS and setting CNAME.

</details>

<details>

<summary>Do I have needs to cooperate?</summary>

- 🟡 Self-hosted services can achieve the same effect as Netlify, but with much more heavy-lifting. Usually, you would have a specific person who looks after the deployment, and the workflow won't be very git-based as opposed to the other two options.

> - 🟢 Netlify and Vercel have deploy previews for every pull request, which is useful for a team to review others' work before doing a production build. You can also manage a team with different access to the deployment.

Reviewer 2: ```suggestion
- 🟢 Netlify and Vercel have deploy previews for every pull request, which is useful for a team to review work before merging to production. You can also manage a team with different access to the deployment.
```
Conciseness

> - 🟡 GitHub pages cannot do deploy previews in a non-convoluted way. One repo can only be associated with one site deployment. On the other hand, you can control who has write access to the site's deployment.

Reviewer 2: ```suggestion
- 🟡 GitHub Pages cannot do deploy previews in a non-convoluted way. One repo can only be associated with one site deployment. On the other hand, you can control who has write access to the site's deployment.
```

</details>

> There isn't a silver bullet. You need to weigh your needs and resources before making your choice.

Reviewer 2: ```suggestion
```
Personally, I would remove this sentence, it just seems out of place in the context of the doc.

Author: I'm trying to stay value-neutral. The color-coding, if you add them up, already seems like I'm biasing towards Netlify which I'm trying hard not to. But yes I see there could be a tone inconsistency

Reviewer 2: I see your point! However, there might be a better way to phrase this.

Author: This section in itself is a bit of a misfit (see also https://github.com/facebook/docusaurus/issues/2953#issuecomment-974631173) I'd say that it's okay to be opinionated here

Reviewer 2: Yeah, we should defer decisions to @slorber.

While you set up the site, specify the build commands and directories as follows:

- build command: `npm run build`
- build directory: `build`

If you did not configure these build options, you may still go to "Site settings" -> "Build and deploy" after your site is created.

Once properly configured with the above options, your site should deploy and automatically redeploy upon merging to your deploy branch, which defaults to `main`.

:::caution

Some Docusaurus sites put the `docs` folder outside of `website` (most likely former Docusaurus v1 sites):

```bash
repo           # git root
├── docs       # md files
└── website    # docusaurus root
```

If you decide to use the `website` folder as Netlify's base directory, Netlify will not trigger builds when you update the `docs` folder, and you need to configure a [custom `ignore` command](https://docs.netlify.com/configure-builds/common-configurations/ignore-builds/):

```toml title="website/netlify.toml"
[build]
  ignore = "git diff --quiet $CACHED_COMMIT_REF $COMMIT_REF . ../docs/"
```

:::

:::warning

By default, Netlify adds trailing slashes to Docusaurus URLs.

> It is recommended to disable the Netlify setting `Post Processing > Asset Optimization > Pretty Urls` to prevent lowercased URLs, unnecessary redirects and 404 errors.

Reviewer 2: ```suggestion
It is recommended to disable the Netlify setting `Post Processing > Asset Optimization > Pretty Urls` to prevent lowercased URLs, unnecessary redirects, and 404 errors.
```

Author: Haha, that's Oxford comma. I'm with you here.

Reviewer 2: :)

**Be very careful**: the `Disable asset optimization` global checkbox is broken and does not really disable the `Pretty URLs` setting in practice. Please make sure to **uncheck it independently**.

If you want to keep the `Pretty Urls` Netlify setting on, adjust the `trailingSlash` Docusaurus config appropriately.

Refer to [slorber/trailing-slash-guide](https://github.com/slorber/trailing-slash-guide) for more information.

:::

## Deploying to Vercel {#deploying-to-vercel}

Deploying your Docusaurus project to [Vercel](https://vercel.com/) will provide you with [various benefits](https://vercel.com/) in the areas of performance and ease of use.

To deploy your Docusaurus project with a [Vercel for Git Integration](https://vercel.com/docs/git-integrations), make sure it has been pushed to a Git repository.

Import the project into Vercel using the [Import Flow](https://vercel.com/import/git). During the import, you will find all relevant options preconfigured for you; however, you can choose to change any of these options, a list of which can be found [here](https://vercel.com/docs/build-step#build-&-development-settings).

> After your project has been imported, all subsequent pushes to branches will generate [Preview Deployments](https://vercel.com/docs/platform/deployments#preview), and all changes made to the [Production Branch](https://vercel.com/docs/git-integrations#production-branch) (commonly "main") will result in a [Production Deployment](https://vercel.com/docs/platform/deployments#production).

Reviewer 2: ```suggestion
After your project has been imported, all subsequent pushes to branches will generate [Preview Deployments](https://vercel.com/docs/platform/deployments#preview), and all changes made to the [Production Branch](https://vercel.com/docs/git-integrations#production-branch) (usually "main") will result in a [Production Deployment](https://vercel.com/docs/platform/deployments#production).
```

> 

Reviewer: worth adding a section about build optimizations, ideally to all hosting providers (but can be done in another PR as it involves some research?)

```
    "netlify-plugin-cache": "^1.0.3",
```

```
[[plugins]]
package = "netlify-plugin-cache"
  [plugins.inputs]
  paths = [
    "node_modules/.cache/webpack",
  ]
```

Author: Yeah, let's add that in another PR

> Now, whenever you push a new tag to github, this trigger will start the drone ci job to publish your website.

Reviewer 2: ```suggestion
Now, whenever you push a new tag to GitHub, this trigger will start the drone CI job to publish your website.
```
